### PR TITLE
CCN3 786 App Login

### PR DESCRIPTION
**Pull request by matthewstyles**
## Todo's before PR

- [x] Rebase from develop
- [x] Unit tests passed (`make test`)
- [x] PR naming according our [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Commits-guideline)
- [x] PR description describes the overall goals of PR commits
- [x] Templates been checked with relevant user-story

## PR Todo's

Please refer to PR [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-submit-a-pull-request)

- [x] Specified reviewers (even if it's yourself)
- [x] Specified assigneses (those who'll merge)
- [x] Specified labels
- [x] Specified milestone

## PR review

Please refer to PR review [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-review-a-pull_request)

- [ ] Code syntax formatting checked
- [ ] Debug messages are absent

## PR merge

Select only one:

- [ ] Should be merged?
- [x] Should be rebased?
- [ ] Should be squashed?

Please refer to PR merge [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-merge-a-pull_request)

### CODE REVIEW

**Comment by cruddasj** [comment]
run jenkins build

File: application/forms/form.py [hunk 1]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+class SecurityQuestionForm(ChildminderForms):
     """
     GOV.UK form for the page to verify an SMS code
     """
     field_label_classes = 'form-label-bold'
     error_summary_template_name = 'error-summary.html'
     auto_replace_widgets = True
-
-    security_answer = forms.CharField(label='Security question', required=True)
+    answer = None
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Would be good to put in a comment/clarification on what this is used for

File: application/forms/form.py [hunk 2]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+        self.answer = kwargs.pop('answer')
+        super(SecurityQuestionForm, self).__init__(*args, **kwargs)
         full_stop_stripper(self)
 
     def clean_security_answer(self):
         security_answer = self.cleaned_data['security_answer']
-        if UserDetails.objects.filter(security_answer=security_answer).count() == 0:
-            raise forms.ValidationError('TBC')
+        if len(security_answer) == 0:
+            print("EMPTY")
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Remove debug statement

File: application/forms/form.py [hunk 3]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+            self.error = 'emtpy'
+            raise forms.ValidationError('empty')
+
+        if self.answer.replace(' ','') != security_answer.replace(' ',''):
+            print("WRONG")
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Remove debug statement

File: application/forms/form.py [hunk 4]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+            self.error = 'wrong'
+            raise forms.ValidationError('wrong')
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Change if debug text, apologies if not

File: application/forms/form.py [hunk 5]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+
+    def clean_security_answer(self):
+        """
+        Date of birth validation (calculate if age is less than 18)
+        :return: birth day, birth month, birth year
+        """
+        birth_day = str(self.cleaned_data['date_of_birth'].day)
+        birth_month = str(self.cleaned_data['date_of_birth'].month)
+        birth_year = str(self.cleaned_data['date_of_birth'].year)
+        if len(str(birth_year)) == 0 or len(str(birth_day)) == 0 or len(str(birth_month)) == 0:
```
Comment (GeevesH-IS) [inline application/forms/form.py]: You can remove the string casting here, as it is done in your assignments (398, 399, 400)

File: application/forms/form.py [hunk 6]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+            print("EMPTYD")
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Remove debug prints

File: application/forms/form.py [hunk 7]
```diff
@@ -354,33 +334,83 @@ def clean_magic_link_sms(self):
...
+            self.error = 'emtpyd'
+            raise forms.ValidationError('empty')
+
+        if self.day != birth_day or self.month != birth_month or self.year != birth_year:
+            print("WRONGD")
+            self.error = 'wrongd'
```
Comment (GeevesH-IS) [inline application/forms/form.py]: Remove Debug Comments

File: application/magic_link.py [hunk 8]
```diff
@@ -145,8 +146,15 @@ def validate_magic_link(request, id):
...
         exp = acc.email_expiry_date
         if not has_expired(exp) and len(id) > 0:
+            if len(acc.mobile_number) == 0:
+                response = HttpResponseRedirect(reverse('Contact-Phone-View')+'?id='+str(app_id))
+                CustomAuthenticationHandler.create_session(response, acc.email)
+                return response
+
             acc.email_expiry_date = 0
             phone = acc.mobile_number
             g = generate_random(5, 'code')
```
Comment (GeevesH-IS) [inline application/magic_link.py]: Probably best to give this a better name

File: application/security_question.py [hunk 9]
```diff
@@ -7,47 +7,171 @@
...
 
-                # Forward back onto application
-                return response
+            # Forward back onto application
+            return response
         else:
+            if 'wrong' in forms:
+                error_message = 'Your answer must match what you told us in your application'
+            else:
+                print(forms)
```
Comment (GeevesH-IS) [inline application/security_question.py]: Remove Debug Message

File: application/security_question.py [hunk 10]
```diff
@@ -7,47 +7,171 @@
...
+        form_list.append(SecurityDateForm(r, day=day, month=month, year=year))
+    if 'oldest' in question:
+        day = date_answer[0]
+        month = date_answer[1]
+        year = date_answer[2]
+        form_list.append(SecurityDateForm(r, day=day, month=month, year=year))
+    if 'dbs' in question:
+        form_list.append(SecurityQuestionForm(r, answer=field_answer))
+    try:
+        for i in form_list:
```
Comment (GeevesH-IS) [inline application/security_question.py]: Would be good to give this a more descriptive name

File: application/security_question.py [hunk 11]
```diff
@@ -7,47 +7,171 @@
...
+            if i.is_valid():
+                i.clean_security_answer()
+            elif i.error == 'wrong':
+                return i.error
+            elif i.error == 'empty':
+                return i.error
+        return True
+    except Exception as ex:
+        print("EX: " + str(ex))
```
Comment (GeevesH-IS) [inline application/security_question.py]: Remove Print Statement

File: application/security_question.py (hunk 10 above)
Comment (GeevesH-IS) [reply application/security_question.py]: The i that is, don't know why its quoted all those lines above it

File: application/magic_link.py (hunk 8 above)
Comment (GeevesH-IS) [reply application/magic_link.py]: The g that is, don't know why its quoted all those lines above it

**Pull request merged**